## Playing around with datasets, part 1:

In [1]:
# import libraries
import os

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
dt_lst = ['EXTR_RPSale.csv', 'EXTR_ResBldg.csv', 'EXTR_Parcel.csv', 'EXTR_LookUp.csv']
dfs = {}

In [3]:
path_to_raw = os.path.join('..', '..', 'data', 'raw')
for f in dt_lst:
    f_key = f[5:-4]
    dfs[f_key] = pd.read_csv(os.path.join(path_to_raw, f), dtype='str')

### RPSale

In [4]:
dfs['RPSale'].head()

ExciseTaxNbr   Major Minor DocumentDate SalePrice    RecordingNbr Volume  \
0      2857854  198920  1430   03/28/2017         0  20170410000541          
1      2743355  638580  0110   07/14/2015    190000  20150715002686          
2      2999169  919715  0200   07/08/2019    192000  20190712001080          
3      2841697  894677  0240   12/21/2016    818161  20161228000896          
4      2826129  445872  0260   10/03/2016         0  20161004000511          

  Page PlatNbr PlatType  ... PropertyType PrincipalUse SaleInstrument  \
0                        ...            3            7             15   
1                        ...            3            6              3   
2                        ...            3            2              3   
3                        ...            2            6              3   
4                        ...            3            2             15   

  AFForestLand AFCurrentUseLand AFNonProfitUse AFHistoricProperty SaleReason  \
0            N                N              N                  N         16   
1            N                N              N                  N          1   
2            N                N              N                  N          1   
3            N                N              N                  N          1   
4            N                N              N                  N         18   

  PropertyClass SaleWarning  
0             2       20 31  
1             8          15  
2             3              
3             8              
4             3       18 31  

[5 rows x 24 columns]

In [5]:
dfs['RPSale'].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 351067 entries, 0 to 351066
Data columns (total 24 columns):
ExciseTaxNbr          351067 non-null object
Major                 351067 non-null object
Minor                 351067 non-null object
DocumentDate          351067 non-null object
SalePrice             351067 non-null object
RecordingNbr          351067 non-null object
Volume                351067 non-null object
Page                  351067 non-null object
PlatNbr               351067 non-null object
PlatType              351067 non-null object
PlatLot               351067 non-null object
PlatBlock             351067 non-null object
SellerName            351067 non-null object
BuyerName             351067 non-null object
PropertyType          351067 non-null object
PrincipalUse          351067 non-null object
SaleInstrument        351067 non-null object
AFForestLand          351067 non-null object
AFCurrentUseLand      351067 non-null object
AFNonProfitUse        351067 non-n

In [6]:
# combine the strings in 'Major' & 'Minor' columns into 1 column
dfs['RPSale']['MM'] = dfs['RPSale'].Major + dfs['RPSale'].Minor
dfs['RPSale'].MM.value_counts()

0660002310    83
2625059212    39
8847500004    38
0425049042    30
3425069018    23
              ..
7298050450     1
2538300060     1
3222059076     1
1320069007     1
2473370550     1
Name: MM, Length: 252962, dtype: int64

In [8]:
# dropping 'Major' & 'Minor' columns
dfs['RPSale'].drop(['Major', 'Minor'], axis=1, inplace=True)

In [9]:
# Check null values
dfs['RPSale'].isna().sum()

ExciseTaxNbr          0
DocumentDate          0
SalePrice             0
RecordingNbr          0
Volume                0
Page                  0
PlatNbr               0
PlatType              0
PlatLot               0
PlatBlock             0
SellerName            0
BuyerName             0
PropertyType          0
PrincipalUse          0
SaleInstrument        0
AFForestLand          0
AFCurrentUseLand      0
AFNonProfitUse        0
AFHistoricProperty    0
SaleReason            0
PropertyClass         0
SaleWarning           0
MM                    0
dtype: int64

### ResBldg

In [10]:
dfs['ResBldg'].head()

Major Minor BldgNbr NbrLivingUnits                     Address  \
0  009800  0720       1              1  27719  SE 26TH WAY   98075   
1  009802  0140       1              1  2829   277TH TER SE  98075   
2  009830  0020       1              1  1715   298TH CRESENT  SE     
3  009830  0160       1              1  1861   297TH WAY SE  98024   
4  010050  0180       1              1    35410   25TH PL S  98003   

  BuildingNumber Fraction DirectionPrefix                 StreetName  \
0          27719                       SE  26TH                        
1          2829                            277TH                       
2          1715                            298TH CRESENT               
3          1861                            297TH                       
4          35410                           25TH                        

  StreetType  ... FpMultiStory FpFreestanding FpAdditional YrBuilt  \
0     WAY     ...            0              0            0    2001   
1     TER     ...            0              0            0    2004   
2             ...            0              0            0    2017   
3     WAY     ...            0              0            0    2013   
4     PL      ...            0              0            0    1994   

  YrRenovated PcntComplete Obsolescence PcntNetCondition Condition AddnlCost  
0           0            0            0                0         3         0  
1           0            0            0                0         3         0  
2           0            0            0                0         3         0  
3           0            0            0                0         3         0  
4           0            0            0                0         4         0  

[5 rows x 50 columns]

In [11]:
# combine the strings in 'Major' & 'Minor' columns into 1 column
dfs['ResBldg']['MM'] = dfs['ResBldg'].Major + dfs['ResBldg'].Minor

In [12]:
# drop 'Major' & 'Minor' columns
dfs['ResBldg'].drop(['Major', 'Minor'], axis=1, inplace=True)

In [13]:
dfs['ResBldg'].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181510 entries, 0 to 181509
Data columns (total 49 columns):
BldgNbr               181510 non-null object
NbrLivingUnits        181510 non-null object
Address               181510 non-null object
BuildingNumber        181510 non-null object
Fraction              181510 non-null object
DirectionPrefix       181146 non-null object
StreetName            181510 non-null object
StreetType            181510 non-null object
DirectionSuffix       181146 non-null object
ZipCode               154594 non-null object
Stories               181510 non-null object
BldgGrade             181510 non-null object
BldgGradeVar          181510 non-null object
SqFt1stFloor          181510 non-null object
SqFtHalfFloor         181510 non-null object
SqFt2ndFloor          181510 non-null object
SqFtUpperFloor        181510 non-null object
SqFtUnfinFull         181510 non-null object
SqFtUnfinHalf         181510 non-null object
SqFtTotLiving         181510 non-n

In [14]:
# Check Null values
dfs['ResBldg'].isna().sum().sort_values(ascending=False)[:5]

ZipCode            26916
DirectionPrefix      364
DirectionSuffix      364
MM                     0
SqFt1stFloor           0
dtype: int64

In [15]:
# identify buildings with Forced-Air heating system (value == 5)
dfs['ResBldg'][dfs['ResBldg'].HeatSystem == '5'].head()

BldgNbr NbrLivingUnits                     Address BuildingNumber Fraction  \
0       1              1  27719  SE 26TH WAY   98075          27719            
1       1              1  2829   277TH TER SE  98075          2829             
2       1              1  1715   298TH CRESENT  SE            1715             
3       1              1  1861   297TH WAY SE  98024          1861             
4       1              1    35410   25TH PL S  98003          35410            

  DirectionPrefix                 StreetName StreetType DirectionSuffix  \
0              SE  26TH                          WAY                      
1                  277TH                         TER                 SE   
2                  298TH CRESENT                                     SE   
3                  297TH                         WAY                 SE   
4                  25TH                          PL                  S    

  ZipCode  ... FpFreestanding FpAdditional YrBuilt YrRenovated PcntComplete  \
0   98075  ...              0            0    2001           0            0   
1   98075  ...              0            0    2004           0            0   
2     NaN  ...              0            0    2017           0            0   
3   98024  ...              0            0    2013           0            0   
4   98003  ...              0            0    1994           0            0   

  Obsolescence PcntNetCondition Condition AddnlCost          MM  
0            0                0         3         0  0098000720  
1            0                0         3         0  0098020140  
2            0                0         3         0  0098300020  
3            0                0         3         0  0098300160  
4            0                0         4         0  0100500180  

[5 rows x 49 columns]

### Parcel

In [16]:
dfs['Parcel'].head()

Unnamed: 0   Major Minor PropName                            PlatName  \
0          0  807841  0410                      SUMMER RIDGE DIV NO. 02   
1          2  755080  0015           SANDER'S TO GILMAN PK & SALMON BAY   
2          3  888600  0135                           VASHON GARDENS ADD   
3          6  022603  9181                                          NaN   
4          7  229670  0160                               ELDORADO NORTH   

          PlatLot PlatBlock Range Township Section  ... SeismicHazard  \
0  41                           6       25      22  ...             N   
1  3                1           3       25      11  ...             N   
2  21                           3       22       8  ...             N   
3                               3       26       2  ...             N   
4  16                           5       26      19  ...             N   

  LandslideHazard SteepSlopeHazard Stream Wetland SpeciesOfConcern  \
0               N                N      N       N                N   
1               N                N      N       N                N   
2               N                N      N       N                N   
3               N                N      N       N                N   
4               N                N      N       N                N   

  SensitiveAreaTract WaterProblems TranspConcurrency OtherProblems  
0                  N             N                 N             N  
1                  N             N                 N             N  
2                  N             N                 N             N  
3                  N             N                 N             N  
4                  N             N                 N             N  

[5 rows x 82 columns]

In [17]:
# combine the strings in 'Major' & 'Minor' columns into 1 column
dfs['Parcel']['MM'] = dfs['Parcel'].Major + dfs['Parcel'].Minor

In [18]:
# drop 'Major' & 'Minor' columns
dfs['Parcel'].drop(['Major', 'Minor'], axis=1, inplace=True)

In [19]:
# Check null values
dfs['Parcel'].isna().sum().sort_values(ascending=False)[:10]

SpecArea              200335
SpecSubArea           200335
PlatName               28545
PropName                9111
Area                       6
SubArea                    6
Topography                 0
StreetSurface              0
RestrictiveSzShape         0
InadequateParking          0
dtype: int64

2 columns `SpecSubArea` and `SpecArea` in the `Parcel` dataframe have very high amount of Null values 
--> might want to consider dropping these.

In [20]:
dfs['Parcel'].drop(['Unnamed: 0', 'SpecSubArea', 'SpecArea'], axis=1, inplace=True)

In [21]:
# Check null values again
dfs['Parcel'].isna().sum().sort_values(ascending=False)[:10]

PlatName               28545
PropName                9111
SubArea                    6
Area                       6
SmallLakeRiverCreek        0
Olympics                   0
StreetSurface              0
RestrictiveSzShape         0
InadequateParking          0
PcntUnusable               0
dtype: int64

In [22]:
dfs['Parcel'].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205199 entries, 0 to 205198
Data columns (total 78 columns):
PropName                  196088 non-null object
PlatName                  176654 non-null object
PlatLot                   205199 non-null object
PlatBlock                 205199 non-null object
Range                     205199 non-null object
Township                  205199 non-null object
Section                   205199 non-null object
QuarterSection            205199 non-null object
PropType                  205199 non-null object
Area                      205193 non-null object
SubArea                   205193 non-null object
DistrictName              205199 non-null object
LevyCode                  205199 non-null object
CurrentZoning             205199 non-null object
HBUAsIfVacant             205199 non-null object
HBUAsImproved             205199 non-null object
PresentUse                205199 non-null object
SqFtLot                   205199 non-null object
WaterSy

### LookUp

In [23]:
dfs['LookUp'].head()

LUType LUItem                                      LUDescription
0    1      1    LAND ONLY                                     ...
1    1      10   Land with new building                        ...
2    1      11   Household, single family units                ...
3    1      12   Multiple family residence (Residential, 2-4 un...
4    1      13   Multiple family residence (Residential, 5+ uni...

In [24]:
# HeatSystem LookUp Type Code is 108
dfs['LookUp'][dfs['LookUp'].LUType == '108']

LUType LUItem                                      LUDescription
243    108    1    Floor-Wall                                    ...
244    108    2    Gravity                                       ...
245    108    3    Radiant                                       ...
246    108    4    Elec BB                                       ...
247    108    5    Forced Air                                    ...
248    108    6    Hot Water                                     ...
249    108    7    Heat Pump                                     ...
250    108    8    Other                                         ...

## Merging datasets

### `dfs['ResBldg']` & `dfs['Parcel']`

In [25]:
# Merging dfs['ResBldg'] & dfs['Parcel'] at 'MM'
# dropping 'Major' & 'Minor' columns to avoid problems with duplicated column names
# their information has been combined into 'MM' anyway
resbldg_parcel = pd.merge(dfs['ResBldg'], dfs['Parcel'], 
                          on='MM', how='inner')

In [26]:
resbldg_parcel.isna().sum().sort_values(ascending=False)[:10]

ZipCode            26916
PlatName           20556
PropName            8071
DirectionPrefix      364
DirectionSuffix      364
FpSingleStory          0
BathFullCount          0
FpMultiStory           0
FpFreestanding         0
Condition              0
dtype: int64

### `dfs['RPSale']` & `resbldg_parcel`

In [27]:
# merge dfs['RPSale'] with resbldg_parcel on 'MM'
df = pd.merge(dfs['RPSale'], resbldg_parcel, on='MM', how='inner')
df.head()

ExciseTaxNbr DocumentDate SalePrice    RecordingNbr Volume Page PlatNbr  \
0      2743355   07/14/2015    190000  20150715002686                       
1      2743356   07/14/2015         0  20150715002687                       
2      2772937   12/21/2015    338000  20151222002066                       
3      2841697   12/21/2016    818161  20161228000896                       
4      3024468   12/05/2019         0  20191209000162                       

  PlatType       PlatLot_x PlatBlock_x  ... SeismicHazard LandslideHazard  \
0                                       ...             N               N   
1                                       ...             N               N   
2                                       ...             N               N   
3                                       ...             N               N   
4                                       ...             N               N   

  SteepSlopeHazard Stream Wetland SpeciesOfConcern SensitiveAreaTract  \
0                N      N       N                N                  N   
1                N      N       N                N                  N   
2                N      N       N                N                  N   
3                N      N       N                N                  N   
4                N      N       N                N                  N   

  WaterProblems TranspConcurrency OtherProblems  
0             N                 N             N  
1             N                 N             N  
2             N                 N             N  
3             N                 N             N  
4             N                 N             N  

[5 rows x 148 columns]

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 251300 entries, 0 to 251299
Columns: 148 entries, ExciseTaxNbr to OtherProblems
dtypes: object(148)
memory usage: 285.7+ MB


In [29]:
df.isna().sum().sort_values(ascending=False)[:6]

ZipCode               39980
PlatName              29223
PropName              11473
DirectionSuffix         604
DirectionPrefix         604
SqFtGarageBasement        0
dtype: int64

A number of columns, such as `Volume`, `Page`, etc. seem to only contain a string of white spaces `'   '`. We might need to identify these and remove them since they don't quite provide any additional information. 

In [30]:
# Remove trailing whitespaces in df
df = df.applymap(lambda x: x.strip() if type(x) == str else x)

In [31]:
# Identify columns with the top value_counts() index being empty string ''
empty_cols = []
for col in df.columns:
    if df[col].value_counts().index[0] == '':
        empty_cols.append(col)
empty_cols

['RecordingNbr',
 'Volume',
 'Page',
 'PlatNbr',
 'PlatType',
 'PlatLot_x',
 'PlatBlock_x',
 'SaleWarning',
 'Address',
 'BuildingNumber',
 'Fraction',
 'DirectionPrefix',
 'DirectionSuffix',
 'PropName',
 'PlatLot_y',
 'PlatBlock_y']

In [32]:
for col in empty_cols:
    print(col)
    print(df[col].value_counts())
    print()

RecordingNbr
                  18638
20191007001027       62
20160104000551       59
20150925000547       54
20170309000731       52
                  ...  
20191108001728        1
20180928001749        1
20170908000024        1
20161128001901        1
20180326001105        1
Name: RecordingNbr, Length: 225202, dtype: int64

Volume
    251300
Name: Volume, dtype: int64

Page
    251300
Name: Page, dtype: int64

PlatNbr
    251300
Name: PlatNbr, dtype: int64

PlatType
    251300
Name: PlatType, dtype: int64

PlatLot_x
    251300
Name: PlatLot_x, dtype: int64

PlatBlock_x
    251300
Name: PlatBlock_x, dtype: int64

SaleWarning
               154990
18 31 51        26085
26               5324
18 31 51 52      5104
18 31 38         4270
                ...  
15 45 51 52         1
13 26 36 45         1
11 18 32 52         1
14 18 32            1
29 31               1
Name: SaleWarning, Length: 1236, dtype: int64

Address
                              649
20904   677TH PL SE  98045    210
18

In [33]:
# Drop 'Volume', 'Page', 'PlatNbr', 'PlatType', 'PlatLot_x', 'PlatBlock_x'
to_drop = ['Volume', 'Page', 'PlatNbr', 'PlatType', 'PlatLot_x', 'PlatBlock_x']
df.drop(to_drop, axis=1, inplace=True)

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 251300 entries, 0 to 251299
Columns: 142 entries, ExciseTaxNbr to OtherProblems
dtypes: object(142)
memory usage: 274.2+ MB


### Converting some columns to Numeric 

#### `SalePrice`

First, we need to convert the `SalePrice` column of `df` into integer

In [35]:
df.SalePrice = df.SalePrice.astype(int)

In [36]:
df.describe()

SalePrice
count  2.513000e+05
mean   5.147809e+05
std    8.367786e+05
min   -6.000000e+02
25%    0.000000e+00
50%    4.142500e+05
75%    7.060000e+05
max    6.250000e+07

In [37]:
df.SalePrice.value_counts()

0          77664
450000      1304
550000      1237
650000      1213
400000      1197
           ...  
413194         1
650750         1
2199026        1
292301         1
866265         1
Name: SalePrice, Length: 21803, dtype: int64

77664 entries in `df` have a `SalePrice` value of 0, which is probably not a good reflection of the value of these houses.

--> These take up roughly 30% of the total number of datapoints we currently have in `df`

--> It might be ok to drop these rows, since including them could potentially skew the results. 

In [39]:
df.head()

ExciseTaxNbr DocumentDate  SalePrice    RecordingNbr            SellerName  \
0      2743355   07/14/2015     190000  20150715002686     GINGRICH AUDREY B   
1      2743356   07/14/2015          0  20150715002687    NATION CHRISTINE M   
2      2772937   12/21/2015     338000  20151222002066         NATION FRED I   
3      2841697   12/21/2016     818161  20161228000896  QUADRANT CORPORATION   
4      3024468   12/05/2019          0  20191209000162             KANG JING   

                                    BuyerName PropertyType PrincipalUse  \
0                               NATION FRED I            3            6   
1                               NATION FRED I            3            6   
2  SUNG NI HLEI+ZING ESTHER ZA+ZOTUNG ZAW MOE            3            6   
3                                   KANG JING            2            6   
4                         KANG JING+XIAO XUAN            3            6   

  SaleInstrument AFForestLand  ... SeismicHazard LandslideHazard  \
0              3            N  ...             N               N   
1             15            N  ...             N               N   
2              3            N  ...             N               N   
3              3            N  ...             N               N   
4             15            N  ...             N               N   

  SteepSlopeHazard Stream Wetland SpeciesOfConcern SensitiveAreaTract  \
0                N      N       N                N                  N   
1                N      N       N                N                  N   
2                N      N       N                N                  N   
3                N      N       N                N                  N   
4                N      N       N                N                  N   

  WaterProblems TranspConcurrency OtherProblems  
0             N                 N             N  
1             N                 N             N  
2             N                 N             N  
3             N                 N             N  
4             N                 N             N  

[5 rows x 142 columns]

And then we can identify column names that are associated with the business questions we are looking to answer, such as Enclosed Porch, Garage / Bedroom conversion, and Heat(ing) System.

In [53]:
# check df.columns for certain substrings:
substr = ['porch', 'garage', 'bedroom', 'heat']

for c in df.columns:
    for s in substr:
        if s in c.lower():
            print(c)

SqFtGarageBasement
SqFtGarageAttached
SqFtOpenPorch
SqFtEnclosedPorch
HeatSystem
HeatSource
Bedrooms


We'll need to make sure we study the data for these columns thoroughly, and also convert them to the appropriate data type.

#### `Bedrooms`

In [43]:
df.Bedrooms.value_counts()

3     104381
4      84410
2      28756
5      24236
6       3938
1       3715
0        749
7        631
8        270
9         96
10        45
14        39
11        15
12         6
13         6
31         2
24         2
19         1
16         1
15         1
Name: Bedrooms, dtype: int64

In [45]:
df.Bedrooms = df.Bedrooms.astype(int)

#### `SqFtEnclosedPorch`

In [51]:
df.SqFtEnclosedPorch.value_counts()

0       239422
40        1210
60        1119
50        1078
70         946
         ...  
204          1
1050         1
25           1
2            1
965          1
Name: SqFtEnclosedPorch, Length: 225, dtype: int64